File: arduino/Stepper-motors/src/helpers/angle.cpp
#include <math.h>

#include "angle.h"

Angle::Angle()
{
  mAngle = 0;
  adjust();
}

Angle::Angle(double iAngle)
{
  mAngle = iAngle;
  adjust();
}

Angle::Angle(double iX, double iY)
{
  double norme = sqrt(pow(iX, 2) + pow(iY, 2));
  // Code pas logique pour des raisons de précisions
  // Voir Ulysse pour plus de détails
  if(iX!=0 && fabs(iY/iX)<1)
  {
    mAngle = asin(iY/norme);
    if(iX<0)
      mAngle = M_PI - asin(iY/norme);
  }
  else
  {
    mAngle = acos(iX/norme);
    if(iY<0)
      mAngle = - acos(iX/norme) - M_PI/2;
  }
}

Angle::~Angle(){}

void Angle::adjust()
{
  while (mAngle > M_PI)
    mAngle -= 2 * M_PI;
  while (mAngle < -M_PI)
    mAngle += 2 * M_PI;
}

double Angle::toDouble() const{return mAngle;}
double Angle::toPositiveDouble() const
{
  if(mAngle<0)
  {
    return mAngle + 2 * M_PI;
  }
  else
  {
    return mAngle;
  }
}
Angle Angle::operator+(Angle iAngle) {return Angle(mAngle + iAngle.mAngle);}
Angle Angle::operator-(Angle iAngle) {return Angle(mAngle - iAngle.mAngle);}
Angle Angle::operator+(double iAngle) {return Angle(mAngle + iAngle);}
Angle Angle::operator-(double iAngle) {return Angle(mAngle - iAngle);}
void Angle::operator=(Angle iAngle) {mAngle=iAngle.mAngle; adjust();}
void Angle::operator=(double iAngle) {mAngle=iAngle; adjust();}

Angle Angle::computeAngle(double startX, double startY, double endX, double endY)
{
  if(fabs(endX - startX) < 1 && fabs(endY - startY) < 1)
  {
    // Not necessary
    return 0;
  }
  double deltaX = endX - startX;
  double alpha;
  if(deltaX == 0)
  {
    if(endY < endX)
      alpha = 3.0 * M_PI/2.0;
    else
      alpha =  M_PI/2.0;
  }
  else if(deltaX > 0)
    alpha =  atan( (endY - startY)/(endX - startX));
  else
    alpha =  M_PI + atan((endY - startY)/(endX - startX));
  return Angle(alpha);
}

File: potato/src/path/angle.py
from math import pi, fabs, atan
from typing import Self

class Angle:
    def __init__(self, angle: float):
        self._angle = angle
        self._adjust()

    def _adjust(self):
        while self._angle > pi:
            self._angle -= 2 * pi
        while self._angle < -pi:
            self._angle += 2 * pi

    def to_float(self):
        self._adjust()
        return self._angle
    
    def to_positive_float(self):
        self._adjust()
        if(self._angle < 0):
            return self._angle + 2 * pi
        else:
            return self._angle
        
    def __add__(self, other: Self) -> Self:
        return Angle(self._angle + other._angle)
    
    def __iadd__(self, other: float) -> Self:
        return Angle(self._angle + other)
    
    def __sub__(self, other: Self) -> Self:
        return Angle(self._angle - other._angle)
    
    def __isub__(self, other: float) -> Self:
        return Angle(self._angle - other)
     
    def __eq__(self, other: Self) -> bool:
        return self._angle == other._angle 
    
    def __str__(self) -> str:
        return f"{self._angle} rad"
    
    @classmethod
    def compute_angle(cls, start_x: float, start_y: float, end_x: float, end_y: float) -> Self:
        if fabs(end_x - start_x) < 1 and fabs(end_y - start_y) < 1:
            # Not necessary
            return Angle(0)
        delta_x = end_x - start_x
        if delta_x == 0:
            if end_y < end_x:
                return Angle(3.0 * pi / 2.0)
            else:
                return Angle(pi / 2)
        elif delta_x > 0:
            return Angle(atan((end_y - start_y)/(end_x - start_x)))
        else:
            return Angle(pi + atan((end_y - start_y) / (end_x - start_x)))